File: scripts/dashboard_server.py
#!/usr/bin/env python3
import http.server
import json
import re
from datetime import datetime
from pathlib import Path

ROOT = Path(__file__).parent.parent
PORT = 3737


def today_prefix():
    now = datetime.now()
    return now.strftime("%y%m%d")  # e.g. 260403


def parse_todos(md):
    sections = []
    current = None
    for line in md.splitlines():
        if line.startswith("## "):
            if current:
                sections.append(current)
            current = {"title": line[3:].strip(), "items": []}
        elif current:
            indent = 1 if line.startswith("  - ") else 0
            stripped = line.strip()
            if stripped.startswith("- [ ]"):
                text = re.sub(r"^- \[ \]\s*", "", stripped)
                current["items"].append({"done": False, "text": text, "indent": indent})
            elif stripped.startswith("- [x]"):
                text = re.sub(r"^- \[x\]\s*", "", stripped)
                current["items"].append({"done": True, "text": text, "indent": indent})
            elif stripped.startswith("- ~~") or stripped.startswith("~~"):
                text = stripped.replace("~~", "").replace("- ", "", 1).strip()
                current["items"].append({"done": True, "text": text, "indent": indent})
    if current:
        sections.append(current)
    return sections


WEALTH_ACCOUNTS = ["成長", "健康", "家庭", "技藝", "金錢", "人脈", "體驗", "服務"]

WEALTH_ALIAS = {
    # 成長
    "成長": "成長", "learning": "成長", "學習": "成長", "growth": "成長",
    # 健康
    "健康": "健康", "health": "健康", "wellness": "健康",
    # 家庭
    "家庭": "家庭", "family": "家庭",
    # 技藝
    "技藝": "技藝", "craft": "技藝", "事業": "技藝", "career": "技藝",
    # 金錢
    "金錢": "金錢", "wealth": "金錢", "財富": "金錢", "money": "金錢",
    # 人脈
    "人脈": "人脈", "relationships": "人脈",
    # 體驗
    "體驗": "體驗", "leisure": "體驗", "休閒": "體驗", "experience": "體驗",
    # 服務
    "服務": "服務", "spirituality": "服務", "心靈": "服務", "service": "服務",
}

def parse_wealth_from_md(md):
    """Extract wealth entries from a summary markdown. Returns dict {account: points}."""
    result = {}
    in_section = False
    for line in md.splitlines():
        if "財富帳戶" in line and line.startswith("##"):
            in_section = True
            continue
        if in_section and line.startswith("##"):
            break
        if in_section and "✦" in line:
            stars = line.count("✦")
            # Extract account name from bold text **帳戶（...）**
            m = re.search(r"\*\*(.+?)[\(（]", line)
            if not m:
                m = re.search(r"\*\*(.+?)\*\*", line)
            if m:
                raw = m.group(1).strip()
                key = WEALTH_ALIAS.get(raw, WEALTH_ALIAS.get(raw.lower()))
                if key:
                    result[key] = result.get(key, 0) + stars
    return result


def get_all_wealth_scores():
    """Scan all summary files and accumulate total + today scores."""
    d = ROOT / "F｜行動聚焦漏斗" / "對話摘要"
    prefix = today_prefix()
    total = {a: 0 for a in WEALTH_ACCOUNTS}
    today = {a: 0 for a in WEALTH_ACCOUNTS}
    if not d.exists():
        return total, today
    for f in sorted(d.iterdir()):
        if not f.suffix == ".md":
            continue
        try:
            md = f.read_text(encoding="utf-8")
            scores = parse_wealth_from_md(md)
            for acc, pts in scores.items():
                if acc in total:
                    total[acc] += pts
                    if f.name.startswith(prefix):
                        today[acc] += pts
        except Exception:
            pass
    return total, today


def get_today_summary():
    prefix = today_prefix()
    d = ROOT / "F｜行動聚焦漏斗" / "對話摘要"
    if not d.exists():
        return None
    files = sorted([f for f in d.iterdir() if f.name.startswith(prefix)])
    if not files:
        return None
    return files[-1].read_text(encoding="utf-8")


def get_manual_notes():
    f = ROOT / "F｜行動聚焦漏斗" / "manual_notes.json"
    if not f.exists():
        return []
    try:
        data = json.loads(f.read_text(encoding="utf-8"))
        return data.get(today_prefix(), [])
    except Exception:
        return []


def save_manual_note(note):
    f = ROOT / "F｜行動聚焦漏斗" / "manual_notes.json"
    data = {}
    if f.exists():
        try:
            data = json.loads(f.read_text(encoding="utf-8"))
        except Exception:
            pass
    key = today_prefix()
    if key not in data:
        data[key] = []
    data[key].insert(0, {
        "text": note,
        "time": datetime.now().strftime("%H:%M")
    })
    f.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")


class Handler(http.server.BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        pass  # quiet

    def send_cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_cors()
        self.end_headers()

    def do_GET(self):
        if self.path in ("/", "/index.html"):
            html = (Path(__file__).parent / "dashboard.html").read_bytes()
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.send_cors()
            self.end_headers()
            self.wfile.write(html)
            return

        if self.path == "/api/data":
            todo_file = ROOT / "F｜行動聚焦漏斗" / "玩家待辦任務.md"
            md = todo_file.read_text(encoding="utf-8")
            now = datetime.now()
            weekdays = ["一", "二", "三", "四", "五", "六", "日"]
            wday = weekdays[now.weekday()]
            date_str = f"{now.year} 年 {now.month} 月 {now.day} 日（週{wday}）"
            wealth_total, wealth_today = get_all_wealth_scores()
            payload = {
                "todos": parse_todos(md),
                "summary": get_today_summary(),
                "notes": get_manual_notes(),
                "date": date_str,
                "wealth_total": wealth_total,
                "wealth_today": wealth_today,
            }
            body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
            self.send_response(200)
            self.send_header("Content-Type", "application/json; charset=utf-8")
            self.send_cors()
            self.end_headers()
            self.wfile.write(body)
            return

        self.send_response(404)
        self.end_headers()

    def do_POST(self):
        if self.path == "/api/note":
            length = int(self.headers.get("Content-Length", 0))
            raw = self.rfile.read(length)
            try:
                data = json.loads(raw)
                note = data.get("note", "").strip()
                if note:
                    save_manual_note(note)
            except Exception:
                pass
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_cors()
            self.end_headers()
            self.wfile.write(b'{"ok":true}')
            return

        self.send_response(404)
        self.end_headers()


if __name__ == "__main__":
    server = http.server.HTTPServer(("localhost", PORT), Handler)
    print(f"✅ FlowGPS 日報啟動：http://localhost:{PORT}")
    print("按 Ctrl+C 停止")
    server.serve_forever()

File: scripts/dashboard-server.js
const http = require('http');
const fs = require('fs');
const path = require('path');
const schedule = require('node-schedule');

const ROOT = path.resolve(__dirname, '..');
const PORT = 3737;

function getFlowJournalFile() {
  return path.join(ROOT, 'F｜行動聚焦漏斗', `flow_journal_${getTodayPrefixNow()}.json`);
}

function getTodayPrefixNow() {
  const now = new Date();
  const y = String(now.getFullYear()).slice(2);
  const m = String(now.getMonth() + 1).padStart(2, '0');
  const d = String(now.getDate()).padStart(2, '0');
  return `${y}${m}${d}`;
}

function getTodayPrefix() {
  const now = new Date();
  const y = String(now.getFullYear()).slice(2);
  const m = String(now.getMonth() + 1).padStart(2, '0');
  const d = String(now.getDate()).padStart(2, '0');
  return `${y}${m}${d}`;
}

function parseTodos(md) {
  const sections = [];
  let current = null;
  for (const line of md.split('\n')) {
    if (line.startsWith('## ')) {
      if (current) sections.push(current);
      current = { title: line.replace('## ', '').trim(), items: [] };
    } else if (current && (line.startsWith('- [ ]') || line.startsWith('  - [ ]'))) {
      const indent = line.startsWith('  ') ? 1 : 0;
      current.items.push({ done: false, text: line.replace(/\s*- \[ \]\s*/, '').trim(), indent });
    } else if (current && (line.startsWith('- [x]') || line.startsWith('  - [x]') || line.startsWith('- ~~') || line.startsWith('  - ~~'))) {
      const indent = line.startsWith('  ') ? 1 : 0;
      current.items.push({ done: true, text: line.replace(/\s*- \[x\]\s*|~~|~~/g, '').trim(), indent });
    }
  }
  if (current) sections.push(current);
  return sections;
}

function getTodaySummary() {
  const prefix = getTodayPrefix();
  const dir = path.join(ROOT, 'F｜行動聚焦漏斗', '對話摘要');
  try {
    const files = fs.readdirSync(dir).filter(f => f.startsWith(prefix));
    if (!files.length) return null;
    // get latest file
    const file = files[files.length - 1];
    return fs.readFileSync(path.join(dir, file), 'utf8');
  } catch { return null; }
}

function getAllNotes() {
  const file = path.join(ROOT, 'F｜行動聚焦漏斗', 'manual_notes.json');
  try { return JSON.parse(fs.readFileSync(file, 'utf8')); } catch { return {}; }
}

function getManualNotes(dateKey) {
  const data = getAllNotes();
  const key = dateKey || getTodayPrefix();
  return data[key] || [];
}

function saveManualNote(note, tag) {
  const file = path.join(ROOT, 'F｜行動聚焦漏斗', 'manual_notes.json');
  const today = getTodayPrefix();
  const data = getAllNotes();
  if (!data[today]) data[today] = [];
  const entry = { text: note, time: new Date().toLocaleTimeString('zh-TW') };
  if (tag) entry.tag = tag;
  data[today].unshift(entry);
  fs.writeFileSync(file, JSON.stringify(data, null, 2));
}

function deleteManualNote(dateKey, index) {
  const file = path.join(ROOT, 'F｜行動聚焦漏斗', 'manual_notes.json');
  const data = getAllNotes();
  if (data[dateKey]) {
    data[dateKey].splice(index, 1);
    fs.writeFileSync(file, JSON.stringify(data, null, 2));
  }
}

function updateManualNote(dateKey, index, newText) {
  const file = path.join(ROOT, 'F｜行動聚焦漏斗', 'manual_notes.json');
  const data = getAllNotes();
  if (data[dateKey] && data[dateKey][index]) {
    data[dateKey][index].text = newText;
    fs.writeFileSync(file, JSON.stringify(data, null, 2));
  }
}

const server = http.createServer((req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') { res.writeHead(204); res.end(); return; }

  if (req.url === '/' || req.url === '/index.html') {
    const html = fs.readFileSync(path.join(__dirname, 'dashboard.html'), 'utf8');
    res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
    res.end(html);
    return;
  }

  const url = new URL(req.url, `http://localhost`);

  if (url.pathname === '/api/data') {
    const dateKey = url.searchParams.get('date') || getTodayPrefix();
    const todoMd = fs.readFileSync(path.join(ROOT, 'F｜行動聚焦漏斗', '玩家待辦任務.md'), 'utf8');
    const allNotes = getAllNotes();
    const payload = {
      todos: parseTodos(todoMd),
      summary: getTodaySummary(),
      notes: getManualNotes(dateKey),
      noteDates: Object.keys(allNotes).sort().reverse(),
      currentDate: dateKey,
      date: new Date().toLocaleDateString('zh-TW', { year: 'numeric', month: 'long', day: 'numeric', weekday: 'long' }),
    };
    res.writeHead(200, { 'Content-Type': 'application/json; charset=utf-8' });
    res.end(JSON.stringify(payload));
    return;
  }

  if (url.pathname === '/api/note' && req.method === 'POST') {
    let body = '';
    req.on('data', c => body += c);
    req.on('end', () => {
      try {
        const { note, tag } = JSON.parse(body);
        if (note && note.trim()) saveManualNote(note.trim(), tag || '');
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ ok: true }));
      } catch { res.writeHead(400); res.end(); }
    });
    return;
  }

  if (url.pathname === '/api/note/delete' && req.method === 'POST') {
    let body = '';
    req.on('data', c => body += c);
    req.on('end', () => {
      try {
        const { date, index } = JSON.parse(body);
        deleteManualNote(date, index);
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ ok: true }));
      } catch { res.writeHead(400); res.end(); }
    });
    return;
  }

  if (url.pathname === '/api/note/update' && req.method === 'POST') {
    let body = '';
    req.on('data', c => body += c);
    req.on('end', () => {
      try {
        const { date, index, text } = JSON.parse(body);
        if (text && text.trim()) updateManualNote(date, index, text.trim());
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ ok: true }));
      } catch { res.writeHead(400); res.end(); }
    });
    return;
  }

  if (url.pathname === '/api/flow-journal' && req.method === 'GET') {
    const dateKey = url.searchParams.get('date') || getTodayPrefixNow();
    const file = path.join(ROOT, 'F｜行動聚焦漏斗', `flow_journal_${dateKey}.json`);
    // 列出所有 flow_journal 日期
    const dir = path.join(ROOT, 'F｜行動聚焦漏斗');
    const dates = fs.readdirSync(dir)
      .filter(f => f.startsWith('flow_journal_') && f.endsWith('.json'))
      .map(f => f.replace('flow_journal_', '').replace('.json', ''))
      .sort().reverse();
    if (fs.existsSync(file)) {
      const journal = JSON.parse(fs.readFileSync(file, 'utf8'));
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ journal, dates, currentDate: dateKey }));
    } else {
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ journal: null, dates, currentDate: dateKey }));
    }
    return;
  }

  if (url.pathname === '/api/flow-journal' && req.method === 'POST') {
    let body = '';
    req.on('data', c => body += c);
    req.on('end', () => {
      try {
        const { journal } = JSON.parse(body);
        fs.writeFileSync(getFlowJournalFile(), JSON.stringify(journal, null, 2));
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ ok: true }));
      } catch { res.writeHead(400); res.end(); }
    });
    return;
  }

  if (url.pathname === '/api/work-okr' && req.method === 'GET') {
    const file = path.join(ROOT, 'work-okr.json');
    try {
      const data = JSON.parse(fs.readFileSync(file, 'utf8'));
      res.writeHead(200, { 'Content-Type': 'application/json; charset=utf-8' });
      res.end(JSON.stringify(data));
    } catch {
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ okr: '', healthMetrics: '', future: '' }));
    }
    return;
  }

  if (url.pathname === '/api/work-okr' && req.method === 'POST') {
    let body = '';
    req.on('data', c => body += c);
    req.on('end', () => {
      try {
        const data = JSON.parse(body);
        data.updatedAt = new Date().toISOString().slice(0, 10);
        fs.writeFileSync(path.join(ROOT, 'work-okr.json'), JSON.stringify(data, null, 2));
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ ok: true }));
      } catch { res.writeHead(400); res.end(); }
    });
    return;
  }

  res.writeHead(404); res.end();
});

server.listen(PORT, () => {
  console.log(`✅ FlowGPS 日報啟動：http://localhost:${PORT}`);
});

// 每天 00:00 自動建立空白日誌
schedule.scheduleJob('0 0 * * *', () => {
  const file = getFlowJournalFile();
  if (!fs.existsSync(file)) {
    const blank = { boss: '', idea: '', f: '', l: '', o: '', w: '', drip: [], review: '' };
    fs.writeFileSync(file, JSON.stringify(blank, null, 2));
    console.log(`✅ 新的一天，日誌已重置：${file}`);
  }
});
